pkg/optimizer/flatzinc_model.go

```diff
@@ -0,0 +1,217 @@
+	}
+	return fmt.Sprintf("array [1..%d] of %s: %s = %s;\n", fzp.NumInstances, fzp.Type, fzp.Name, fzp.Assignment)
+}
+
+type Annotations []string
+
+func (annots Annotations) AnnotationString() string {
+	annotsStr := strings.Join(annots, " :: ")
+	if len(annotsStr) > 0 {
+		annotsStr = " :: " + annotsStr
+	}
+	return annotsStr
+}
+
+type FlatZincVariable struct {
+	NumInstances uint // NumInstances > 1 means an array of vars
+	Name         string
+	Type         string
+	Assignment   string
+	Annotations
+}
+
+func (fzv FlatZincVariable) VarDeclaration() string {
+	if fzv.NumInstances == 1 {
+		return fmt.Sprintf("var %s: %s%s;\n", fzv.Type, fzv.Name, fzv.AnnotationString())
+	}
+	return fmt.Sprintf("array [1..%d] of var %s: %s%s;\n", fzv.NumInstances, fzv.Type, fzv.Name, fzv.AnnotationString())
+}
+
+type FlatZincConstraint struct {
+	Identifier  string
+	Expressions []string
+	Annotations
+}
+
+func (cnstr FlatZincConstraint) ConstraintStatement() string {
+	exprs := strings.Join(cnstr.Expressions, ", ")
+	return fmt.Sprintf("constraint %s(%s)%s;\n", cnstr.Identifier, exprs, cnstr.AnnotationString())
+}
+
+type SolveGoal int64
+
+const (
+	Satisfy SolveGoal = iota
+	Minimize
+	Maximize
+)
+
+func (s SolveGoal) String() string {
+	switch s {
+	case Satisfy:
+		return "satisfy"
+	case Minimize:
+		return "minimize"
+	case Maximize:
+		return "maximize"
+	}
+	return "unknown"
+}
+
+type FlatZincSolveItem struct {
+	goal SolveGoal
+	expr string
+	Annotations
+}
+
+func (slv FlatZincSolveItem) SolveItemStatement() string {
+	return fmt.Sprintf("solve%s %s %s;\n", slv.AnnotationString(), slv.goal, slv.expr)
+}
+
+type FlatZincModel struct {
+	ParamMap    map[string]FlatZincParam
+	VarMap      map[string]FlatZincVariable
+	Constraints []FlatZincConstraint
+	SolveTarget FlatZincSolveItem
+}
+
+func NewFlatZincModel() *FlatZincModel {
+	var fzw FlatZincModel
+	fzw.ParamMap = make(map[string]FlatZincParam)
+	fzw.VarMap = make(map[string]FlatZincVariable)
+	fzw.Constraints = make([]FlatZincConstraint, 0)
+	return &fzw
+}
+
+func (fzw *FlatZincModel) AddParam(numInst uint, name, vartype, assignment string) {
+	fzw.ParamMap[name] = FlatZincParam{numInst, name, vartype, assignment}
+}
+
+func (fzw *FlatZincModel) AddVariable(numInst uint, name, vartype, assignment string, annotations ...string) {
+	fzw.VarMap[name] = FlatZincVariable{numInst, name, vartype, assignment, annotations}
+}
+
+func (fzw *FlatZincModel) AddConstraint(identifier string, exprs []string, annotations ...string) {
+	fzw.Constraints = append(fzw.Constraints, FlatZincConstraint{identifier, exprs, annotations})
+}
+
+func (fzw *FlatZincModel) SetSolveTarget(goal SolveGoal, expr string, annotations ...string) {
+	fzw.SolveTarget = FlatZincSolveItem{goal, expr, annotations}
+}
+
+func (fzw *FlatZincModel) Dump(fileName string) error {
+	file, err := os.Create(path.Clean(fileName))
+	if err != nil {
+		return fmt.Errorf("failed opening file %s for writing", fileName)
```

Comment:
Please use standardized fybrik logging approach.  https://fybrik.io/dev/contribute/logging/.  There are helper functions in pkg/logging

Reply:
@simanadler This is the returned error, the caller function can log it.

Reply:
Consider using errors.Wrapf to preserve the original error details